Don't passing a nil value to `case_sensitive_comparison`

kamipo opened the pull request:
A `value` is only used for checking `value.nil?`. It is not needed if
immediately return when `value.nil?`.

matthewd commented:
This builds on https://github.com/rails/rails/pull/22877, which [seems problematic](https://github.com/rails/rails/pull/22877#issuecomment-168335900).

rafaelfranca commented:
@matthewd better now?

matthewd commented on activerecord/lib/active_record/validations/uniqueness.rb:
```diff
@@ -65,26 +65,24 @@ def build_relation(klass, table, attribute, value) #:nodoc:
-          bind = Relation::QueryAttribute.new(attribute.to_s, value, Type::Value.new)
-          klass.unscoped.where(comparison, bind)
-        end
+        comparison = klass.connection.uniqueness_comparison(
```
Construct the BindParam here, and then pass it as `value`. That way, it gets created in a place where its real value is known, and the relationship between the two is clear.

kamipo replied:
Thank you for the review. I fixed to construct the `Arel::Nodes::BindParam` here.

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -608,11 +608,8 @@ def primary_keys(table_name) # :nodoc:
-        else
-          table[attribute].eq(Arel::Nodes::Bin.new(Arel::Nodes::BindParam.new))
-        end
+        value = Arel::Nodes::Bin.new(value) if column.collation && !column.case_sensitive?
```
Fixed condition to wrap `Arel::Nodes::Bin.new(value)` only string column.

sgrif commented:
I really don't want validations to be coupled to Arel ASTs if possible.

kamipo commented:
I agree too. I think we can move the building Arel ASTs to predicate builder at least.
But currently #23523 is blocker.

kamipo committed:
Don't passing a nil value to `case_sensitive_comparison`

A `value` is only used for checking `value.nil?`. It is unnecessary if
immediately return when `value.nil?`.
